File: Y2/POO/lab11/visitor_ex.cpp
// da, e luat de pe geeksforgeeks, dar am inteles
// il scriu de mana pe cel de kotlin

#include <iostream>
#include <vector>
#include <cmath>

using namespace std;

class Circle;
class Square;
class Triangle;

class ShapeVisitor {
public:
    virtual void visit(Circle& circle) = 0;
    virtual void visit(Square& square) = 0;
    virtual void visit(Triangle& triangle) = 0;
    virtual ~ShapeVisitor() = default;
};

class Shape {
public:
    virtual void accept(ShapeVisitor& visitor) = 0;
    virtual ~Shape() = default;
};

class Circle : public Shape {
public:
    void accept(ShapeVisitor& visitor) override {
        visitor.visit(*this);
    }
};

class Square : public Shape {
public:
    void accept(ShapeVisitor& visitor) override {
        visitor.visit(*this);
    }
};

class Triangle : public Shape {
public:
    void accept(ShapeVisitor& visitor) override {
        visitor.visit(*this);
    }
};

class AreaCalculator : public ShapeVisitor {
private:
    double totalArea = 0.0;

    double radiusOfCircle = 5.0;
    double sideOfSquare = 4.0;
    double baseOfTriangle = 3.0;
    double heightOfTriangle = 6.0;

public:
    void visit(Circle&) override {
        totalArea += M_PI * pow(radiusOfCircle, 2);
    }

    void visit(Square&) override {
        totalArea += pow(sideOfSquare, 2);
    }

    void visit(Triangle&) override {
        totalArea += (baseOfTriangle * heightOfTriangle) / 2;
    }

    double getTotalArea() const {
        return totalArea;
    }
};

// Main
int main() {
    vector<Shape*> shapes;

    shapes.push_back(new Circle());
    shapes.push_back(new Square());
    shapes.push_back(new Triangle());

    AreaCalculator areaCalculator;

    for (Shape* shape : shapes) {
        shape->accept(areaCalculator);
    }

    cout << "Total area: " << areaCalculator.getTotalArea() << endl;

    for (Shape* shape : shapes) {
        delete shape;
    }

    return 0;
}

File: Y2/POO/lab11/kotlin/visitor_ex.kt
import kotlin.math.*

interface Visitor{
    fun visit(circle: Circle): Double
    fun visit(triangle: Triangle): Double
}

interface Shape{
    fun accept(v: Visitor): Double
}

class Circle(
    var radius: Double
): Shape{
    override fun accept(v: Visitor): Double{
        return v.visit(this)
    }
}

class Triangle(
    var side1: Double,
    var side2: Double,
    var base: Double,
): Shape{
    override fun accept(v: Visitor): Double{
        return v.visit(this)
    }

    var height: Double = 0.0

    init{
        val a = side1
        val b = side2
        val c = base
        val s: Double = (a+ b+ c)/2
        val heronFormula: Double = sqrt(s*(s-a)*(s-b)*(s-c))
        height = 2* heronFormula / a
    }
}

class AreaCalculator: Visitor{

    override fun visit(circle: Circle): Double{
        return PI * circle.radius * circle.radius
    }

    override fun visit(triangle: Triangle): Double{
        return  triangle.base * triangle.height / 2
    }
}

class PerimeterCalculator: Visitor{
    override fun visit(circle: Circle): Double{
        return 2 * PI * circle.radius
    }

    override fun visit(triangle: Triangle): Double{
        return triangle.side1 + triangle.base + triangle.side2
    }    
}

fun main(){
    var shapes: MutableList<Shape> = mutableListOf(
        Circle(12.0),
        Triangle(3.0,4.0,5.0)
    )

   println("The area of the cirlce is: ${shapes[0].accept(AreaCalculator())}")

   println("The area of the triangle is: ${shapes[1].accept(AreaCalculator())}")


    println("The circumference of the circle is: ${shapes[0].accept(PerimeterCalculator())} ")

        println("The perimeter of the triangle is: ${shapes[1].accept(PerimeterCalculator())} ")

}